oml/utils/ddp.py

```diff
@@ -0,0 +1,208 @@
+    """
+    Function inspects loader and modifies sampler for working in DDP mode.
+    Note:
+        We ALWAYS use padding of samples (number of batches or number of samples per epoch) in order to use same amount
+        of data for each device in DDP, so behaviour with and without DDP may be slightly different (e.g. metrics).
+    """
+    if is_ddp():
+        kwargs_loader = {
+            "collate_fn": loader.collate_fn,
+            "persistent_workers": loader.persistent_workers,
+            "pin_memory": loader.pin_memory,
+            "worker_init_fn": loader.worker_init_fn,
+            "prefetch_factor": loader.prefetch_factor,
+            "multiprocessing_context": loader.multiprocessing_context,
+            "pin_memory_device": loader.pin_memory_device,
+            "num_workers": loader.num_workers,
+        }
+
+        # If you don't spectify batch_sampler, PyTorch automatically creates default BatchSampler. In this case we
+        # need convert to DDP only sampler (your custom sampler / default SequentialSampler or RandomSampler, which
+        # PyTorch creates if sampler=None). We don't use `isinstance(...)` for `if` statement because we need exactly
+        # class BatchSampler, ignoring any inheritance
+        if type(loader.batch_sampler) is BatchSampler:
+            shuffle = True if isinstance(loader.sampler, RandomSampler) else False
+            ddp_sampler = DDPSamplerWrapper(
+                sampler=loader.sampler, shuffle_samples_between_gpus=shuffle, pad_data_to_num_gpus=True
+            )
+            patched_loader = DataLoader(
+                dataset=loader.dataset,
+                sampler=ddp_sampler,
+                batch_size=loader.batch_size,
+                drop_last=loader.drop_last,
+                **kwargs_loader,
+            )
+            sampler_info = f"'{loader.sampler.__class__.__name__}' sampler"
+        else:
+            ddp_sampler = DDPSamplerWrapper(
+                sampler=loader.batch_sampler, shuffle_samples_between_gpus=True, pad_data_to_num_gpus=True
+            )
+            patched_loader = DataLoader(dataset=loader.dataset, batch_sampler=ddp_sampler, **kwargs_loader)
+            sampler_info = f"'{loader.batch_sampler.__class__.__name__}' batch sampler"
+
+        logging.info(f"DataLoader with {sampler_info} is updated to DDP mode")
+        return patched_loader
+    else:
+        warnings.warn(patch_dataloader_to_ddp.__name__, WarningDDP)
+        return loader
+
+
+def check_loaders_is_patched(loaders: Union[DataLoader, Sequence[DataLoader]]) -> bool:
+    loaders = [loaders] if isinstance(loaders, DataLoader) else loaders
+
+    for loader in loaders:
+        if not any(isinstance(sampler, DDPSamplerWrapper) for sampler in [loader.batch_sampler, loader.sampler]):
+            return False
+
+    return True
+
+
+def is_ddp() -> bool:
+    try:
+        get_world_size()
+        return True
+    except:
```

Comment:
is it possible to specify the exact error?

Reply:
done